### DISCUSSION TITLE
gs_usb implemented with tinyusb

### DISCUSSION THREAD

**Discussion by Daniel-Trevitz** [hathach/tinyusb#1657] [Q&A] [answered]
Hello, I am working on porting candlelight to use tinyusb: https://github.com/Daniel-Trevitz/candleLight_fw/tree/tinyusb
Candlelight takes a USB + CAN STM uC to create low cost can adapters.

I have EP0 and can rx working, however only one CAN tx frame work, corresponding to one transmission from Linux to the uC on EP2.

My tinyusb branch of candlelight uses a vendor device to implement (most) of the gs_usb spec.
Testing is done on the gs_usb linux driver, and a cdc device for debugging.

The specific CPU I am using is the STM32F072C8Tx

Things tried:
1) Removing neo pixel DMA
2) Removing the CAN code (hardware not initialized or accessed)
3) various combinations of the vendord interface
4) Disabling all calls to tud_vendor_n_*()
5) "safe" printf from dcd_ep_ctr_rx_handler dcd_stm32_fsdev.c

When using printf is seems that EP2 is only running in the ISR once.

I am putting this here in case anyone had any ideas, as I am out of them...
I've reviewed issue #900 as it seems very similar, but I'm not seeing much evidence of a race condition.

**Comment by Daniel-Trevitz** [comment]
FYI @marckleinebudde & @fenugrec

**Comment by Daniel-Trevitz** [comment, marked as answer]
Setting

#define CFG_TUD_VENDOR_TX_BUFSIZE 32
#define CFG_TUD_VENDOR_RX_BUFSIZE 32

to

#define CFG_TUD_VENDOR_TX_BUFSIZE 64
#define CFG_TUD_VENDOR_RX_BUFSIZE 64

has fixed the issue.

**Comment by fenugrec** [comment]
> has fixed the issue.

As in "haven't seen the issue again yet on this combination of hardware and OS",  or "definitely fixed the issue with a rational explanation" ? In my limited experience with USB, the first case is the most frequent...

**Reply by Daniel-Trevitz** [reply]
I’ve  done no testing outside my Linux box; however the fix is rational. I’ve not dove deep but I suspect it was memory corruption in the usb silicon.
